lxd/network/network_utils.go

```diff
@@ -150,6 +150,46 @@ func UsedBy(s *state.State, networkProjectName string, networkName string, first
 		}
 	}
 
+	// Only networks defined in the default project can be used by other networks.
+	if networkProjectName == project.Default {
+		// Get a list of projects (including default).
+		var projectNames []string
+
+		err = s.Cluster.Transaction(func(tx *db.ClusterTx) error {
+			projectNames, err = tx.GetProjectNames()
+			return err
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to load projects")
+		}
+
+		for _, projectName := range projectNames {
+			// Look for networks in the project.
+			networkNames, err := s.Cluster.GetNetworks(projectName)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, name := range networkNames {
+				_, network, err := s.Cluster.GetNetworkInAnyState(projectName, name)
+				if err != nil {
+					return nil, err
+				}
+
+				// The network's config references the network we are searching for. Either by
+				// directly referencing our network or by referencing our interface as its parent.
+				if network.Config["network"] == networkName || network.Config["parent"] == networkName {
+					uri := fmt.Sprintf("/%s/networks/%s", version.APIVersion, network.Name)
+					if projectName != project.Default {
+						uri += fmt.Sprintf("?project=%s", projectName)
+					}
+
+					usedBy = append(usedBy, uri)
+				}
+			}
+		}
+	}
+
```

Comment:
I'm not entirely sure how this function is going to be used and if it's going to be performance/scale sensitive. If there are any concerns around that, it'd be probably good to use a single database transaction to fetch all needed data, instead of ~NxM transaction (N number of projects, M number of networks in each project).

It shoudl be possible to extract from `Cluster.GetNetworks` and `Cluster.GetNetworkInAnyState` their `ClusterTx.GetNetworks` and `ClusterTx.GetNetworkInAnyState` equivalent.

Again, this is needed only if there are any concern about performance (e.g. if this function is going to be used multiple times, or in speed-sensitive code paths). If not, then this is just a remark.

Reply:
@freeekanayaka good point. Its only used by `doNetworkGet` which is mostly used in a singular network context (i.e it is only called once). However it is also used in the `lxc networks list` command which means it will be called for every network listed (except loopback). So this may slow down that command a bit.

Also a potential optimisation here is to only perform the used by check for managed networks. However I am not sure what the contract here is, should one expect to get a list of resources that use non-managed network interfaces?

Reply:
@freeekanayaka I've updated this function to use `tx.GetNonPendingNetworks` and modified it to return a map of projects and networks in a single transaction. As this function `tx.GetNonPendingNetworks` was only used in one other place (also inside network package).